builder/dockerfile/command/commands.go

```diff
@@ -0,0 +1,232 @@
+package command
+
+import (
+	"errors"
+
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/strslice"
+)
+
+type KeyValuePair struct {
+	Key   string
+	Value string
+}
+
+type KeyValuePairs []KeyValuePair
+
+// WithSourceCode is a marker indicating that a given command
+// Has been parsed from source code (wich can be displayed in builder output)
+type WithSourceCode interface {
+	SourceCode() string
+}
+
+type EnvCommand struct {
+	OriginalSource string
+	Env            KeyValuePairs // kvp slice instead of map to preserve ordering
+}
+
+func (c *EnvCommand) SourceCode() string {
```

Comment:
Ok for the sourcecode in the embedded struct, but the normalized command name would have to be put in a field and I don't like it. Could'nt we have an interface 
```
type DispatchCommand interface{
 CommandName() string
}
``` 
implemented by each command instead ? (that wouldn't need a field). But anyway I don't see the point here, as dispatchers need to deal with the typed command anyway (so if you switch on a command name instead of command type, you would need to cast it to the concrete type anyway)